File: server/routes/playerStats.js
const express = require('express');
const router = express.Router();
const database = require('../services/database');

// Transform player stats data for API response
const transformPlayerStats = (stats) => ({
  gameId: stats.gameId,
  playerId: stats.playerId,
  teamId: stats.teamId,
  points: stats.points,
  pos: stats.pos,
  min: stats.min,
  fgm: stats.fgm,
  fga: stats.fga,
  fgp: stats.fgp,
  ftm: stats.ftm,
  fta: stats.fta,
  ftp: stats.ftp,
  tpm: stats.tpm,
  tpa: stats.tpa,
  tpp: stats.tpp,
  offReb: stats.offReb,
  defReb: stats.defReb,
  totReb: stats.totReb,
  assists: stats.assists,
  pFouls: stats.pFouls,
  steals: stats.steals,
  turnovers: stats.turnovers,
  blocks: stats.blocks,
  plusMinus: stats.plusMinus,
  comment: stats.comment,
  player: {
    id: stats.playerId,
    displayName: stats.displayName,
    firstName: stats.firstName,
    lastName: stats.lastName,
    jerseyNumber: stats.jerseyNumber,
    position: stats.position
  },
  team: {
    id: stats.teamId,
    name: stats.teamName,
    city: stats.teamCity,
    abbreviation: stats.teamAbbreviation
  },
  game: {
    id: stats.gameId,
    gameDate: stats.gameDate,
    homeTeamId: stats.homeTeamId,
    awayTeamId: stats.awayTeamId
  },
  createdAt: stats.createdAt,
  updatedAt: stats.updatedAt
});

// GET /api/playerStats - Get player stats with query filters
router.get('/', async (req, res) => {
  try {
    const { gameId, playerId, teamId } = req.query;
    
    let stats;
    
    if (gameId && playerId) {
      // Get specific player stats for a game
      const stat = await database.getPlayerStats(gameId, playerId);
      stats = stat ? [stat] : [];
    } else if (gameId) {
      // Get all stats for a game
      stats = await database.getPlayerStatsByGame(gameId);
    } else if (playerId) {
      // Get all stats for a player
      stats = await database.getPlayerStatsByPlayer(playerId);
    } else if (teamId) {
      // Get all stats for a team
      stats = await database.getPlayerStatsByTeam(teamId);
    } else {
      return res.status(400).json({
        success: false,
        message: 'At least one filter parameter (gameId, playerId, teamId) is required'
      });
    }
    
    // Transform data for API response
    const transformedStats = stats.map(transformPlayerStats);
    
    res.json({
      success: true,
      data: transformedStats,
      count: transformedStats.length,
      filters: {
        gameId: gameId || null,
        playerId: playerId || null,
        teamId: teamId || null
      }
    });

  } catch (error) {
    console.error('Error fetching player stats:', error);
    res.status(500).json({
      success: false,
      message: 'Internal server error',
      error: error.message
    });
  }
});

// GET /api/playerStats/game/:gameId - Get all stats for a game
router.get('/game/:gameId', async (req, res) => {
  try {
    const { gameId } = req.params;
    
    const stats = await database.getPlayerStatsByGame(gameId);
    const transformedStats = stats.map(transformPlayerStats);
    
    res.json({
      success: true,
      data: transformedStats,
      count: transformedStats.length,
      gameId: gameId
    });

  } catch (error) {
    console.error('Error fetching game stats:', error);
    res.status(500).json({
      success: false,
      message: 'Internal server error',
      error: error.message
    });
  }
});

// GET /api/playerStats/player/:playerId - Get all stats for a player
router.get('/player/:playerId', async (req, res) => {
  try {
    const { playerId } = req.params;
    
    const stats = await database.getPlayerStatsByPlayer(playerId);
    const transformedStats = stats.map(transformPlayerStats);
    
    res.json({
      success: true,
      data: transformedStats,
      count: transformedStats.length,
      playerId: playerId
    });

  } catch (error) {
    console.error('Error fetching player stats:', error);
    res.status(500).json({
      success: false,
      message: 'Internal server error',
      error: error.message
    });
  }
});

// GET /api/playerStats/team/:teamId - Get all stats for a team
router.get('/team/:teamId', async (req, res) => {
  try {
    const { teamId } = req.params;
    const { gameId } = req.query;
    
    const stats = await database.getPlayerStatsByTeam(teamId, gameId);
    const transformedStats = stats.map(transformPlayerStats);
    
    res.json({
      success: true,
      data: transformedStats,
      count: transformedStats.length,
      teamId: teamId,
      gameId: gameId || null
    });

  } catch (error) {
    console.error('Error fetching team stats:', error);
    res.status(500).json({
      success: false,
      message: 'Internal server error',
      error: error.message
    });
  }
});

// GET /api/playerStats/:gameId/:playerId - Get specific player stats for a game
router.get('/:gameId/:playerId', async (req, res) => {
  try {
    const { gameId, playerId } = req.params;
    
    const stats = await database.getPlayerStats(gameId, playerId);
    
    if (!stats) {
      return res.status(404).json({
        success: false,
        message: 'Player stats not found for this game'
      });
    }
    
    const transformedStats = transformPlayerStats(stats);
    
    res.json({
      success: true,
      data: transformedStats
    });

  } catch (error) {
    console.error('Error fetching player stats:', error);
    res.status(500).json({
      success: false,
      message: 'Internal server error',
      error: error.message
    });
  }
});

// POST /api/playerStats - Create new player stats
router.post('/', async (req, res) => {
  try {
    const statsData = req.body;
    
    // Validate required fields
    const requiredFields = ['gameId', 'playerId', 'teamId'];
    const missingFields = requiredFields.filter(field => !statsData[field]);
    
    if (missingFields.length > 0) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields',
        missingFields: missingFields
      });
    }
    
    await database.savePlayerStats(statsData);
    
    // Fetch the created stats to return
    const stats = await database.getPlayerStats(statsData.gameId, statsData.playerId);
    const transformedStats = transformPlayerStats(stats);
    
    res.status(201).json({
      success: true,
      data: transformedStats,
      message: 'Player stats created successfully'
    });

  } catch (error) {
    console.error('Error creating player stats:', error);
    res.status(500).json({
      success: false,
      message: 'Internal server error',
      error: error.message
    });
  }
});

// PUT /api/playerStats/:gameId/:playerId - Update player stats
router.put('/:gameId/:playerId', async (req, res) => {
  try {
    const { gameId, playerId } = req.params;
    const statsData = req.body;
    
    // Ensure the IDs match the URL parameters
    statsData.gameId = gameId;
    statsData.playerId = playerId;
    
    await database.savePlayerStats(statsData);
    
    // Fetch the updated stats to return
    const stats = await database.getPlayerStats(gameId, playerId);
    const transformedStats = transformPlayerStats(stats);
    
    res.json({
      success: true,
      data: transformedStats,
      message: 'Player stats updated successfully'
    });

  } catch (error) {
    console.error('Error updating player stats:', error);
    res.status(500).json({
      success: false,
      message: 'Internal server error',
      error: error.message
    });
  }
});

// DELETE /api/playerStats/:gameId/:playerId - Delete player stats
router.delete('/:gameId/:playerId', async (req, res) => {
  try {
    const { gameId, playerId } = req.params;
    
    // Check if stats exist
    const existingStats = await database.getPlayerStats(gameId, playerId);
    
    if (!existingStats) {
      return res.status(404).json({
        success: false,
        message: 'Player stats not found for this game'
      });
    }
    
    const deletedCount = await database.deletePlayerStats(gameId, playerId);
    
    res.json({
      success: true,
      message: 'Player stats deleted successfully',
      deletedCount: deletedCount
    });

  } catch (error) {
    console.error('Error deleting player stats:', error);
    res.status(500).json({
      success: false,
      message: 'Internal server error',
      error: error.message
    });
  }
});

module.exports = router;

File: sports/Models/PlayerStats.swift
//
//  PlayerStats.swift
//  sports
//
//  Created by Joyce Zhang on 1/15/25.
//

import Foundation

struct PlayerStats: Identifiable, Codable {
    let id: String
    let gameId: String
    let playerId: String
    let teamId: String
    let points: Int
    let pos: String
    let min: String
    let fgm: Int
    let fga: Int
    let fgp: String
    let ftm: Int
    let fta: Int
    let ftp: String
    let tpm: Int
    let tpa: Int
    let tpp: String
    let offReb: Int
    let defReb: Int
    let totReb: Int
    let assists: Int
    let pFouls: Int
    let steals: Int
    let turnovers: Int
    let blocks: Int
    let plusMinus: String
    let comment: String?
    let player: PlayerInfo
    let team: TeamInfo
    let game: GameInfo
    let createdAt: String?
    let updatedAt: String?
    
    enum CodingKeys: String, CodingKey {
        case gameId, playerId, teamId, points, pos, min
        case fgm, fga, fgp, ftm, fta, ftp, tpm, tpa, tpp
        case offReb, defReb, totReb, assists, pFouls
        case steals, turnovers, blocks, plusMinus, comment
        case player, team, game, createdAt, updatedAt
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        // Generate id from composite key (gameId, playerId) since database uses composite primary key
        gameId = try container.decode(String.self, forKey: .gameId)
        playerId = try container.decode(String.self, forKey: .playerId)
        id = "\(gameId)_\(playerId)"
        teamId = try container.decode(String.self, forKey: .teamId)
        points = try container.decode(Int.self, forKey: .points)
        pos = try container.decode(String.self, forKey: .pos)
        min = try container.decode(String.self, forKey: .min)
        fgm = try container.decode(Int.self, forKey: .fgm)
        fga = try container.decode(Int.self, forKey: .fga)
        fgp = try container.decode(String.self, forKey: .fgp)
        ftm = try container.decode(Int.self, forKey: .ftm)
        fta = try container.decode(Int.self, forKey: .fta)
        ftp = try container.decode(String.self, forKey: .ftp)
        tpm = try container.decode(Int.self, forKey: .tpm)
        tpa = try container.decode(Int.self, forKey: .tpa)
        tpp = try container.decode(String.self, forKey: .tpp)
        offReb = try container.decode(Int.self, forKey: .offReb)
        defReb = try container.decode(Int.self, forKey: .defReb)
        totReb = try container.decode(Int.self, forKey: .totReb)
        assists = try container.decode(Int.self, forKey: .assists)
        pFouls = try container.decode(Int.self, forKey: .pFouls)
        steals = try container.decode(Int.self, forKey: .steals)
        turnovers = try container.decode(Int.self, forKey: .turnovers)
        blocks = try container.decode(Int.self, forKey: .blocks)
        plusMinus = try container.decode(String.self, forKey: .plusMinus)
        comment = try container.decodeIfPresent(String.self, forKey: .comment)
        player = try container.decode(PlayerInfo.self, forKey: .player)
        team = try container.decode(TeamInfo.self, forKey: .team)
        game = try container.decode(GameInfo.self, forKey: .game)
        createdAt = try container.decodeIfPresent(String.self, forKey: .createdAt)
        updatedAt = try container.decodeIfPresent(String.self, forKey: .updatedAt)
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        
        // id is computed from composite key, not encoded
        try container.encode(gameId, forKey: .gameId)
        try container.encode(playerId, forKey: .playerId)
        try container.encode(teamId, forKey: .teamId)
        try container.encode(points, forKey: .points)
        try container.encode(pos, forKey: .pos)
        try container.encode(min, forKey: .min)
        try container.encode(fgm, forKey: .fgm)
        try container.encode(fga, forKey: .fga)
        try container.encode(fgp, forKey: .fgp)
        try container.encode(ftm, forKey: .ftm)
        try container.encode(fta, forKey: .fta)
        try container.encode(ftp, forKey: .ftp)
        try container.encode(tpm, forKey: .tpm)
        try container.encode(tpa, forKey: .tpa)
        try container.encode(tpp, forKey: .tpp)
        try container.encode(offReb, forKey: .offReb)
        try container.encode(defReb, forKey: .defReb)
        try container.encode(totReb, forKey: .totReb)
        try container.encode(assists, forKey: .assists)
        try container.encode(pFouls, forKey: .pFouls)
        try container.encode(steals, forKey: .steals)
        try container.encode(turnovers, forKey: .turnovers)
        try container.encode(blocks, forKey: .blocks)
        try container.encode(plusMinus, forKey: .plusMinus)
        try container.encodeIfPresent(comment, forKey: .comment)
        try container.encode(player, forKey: .player)
        try container.encode(team, forKey: .team)
        try container.encode(game, forKey: .game)
        try container.encodeIfPresent(createdAt, forKey: .createdAt)
        try container.encodeIfPresent(updatedAt, forKey: .updatedAt)
    }
}

struct PlayerInfo: Codable {
    let id: String
    let displayName: String
    let firstName: String
    let lastName: String
    let jerseyNumber: Int?
    let position: String?
}

struct TeamInfo: Codable {
    let id: String
    let name: String
    let city: String
    let abbreviation: String
}

struct GameInfo: Codable {
    let id: String
    let gameTime: String
    let homeTeamId: String
    let awayTeamId: String
}

// MARK: - Response Models
struct YourServerPlayerStatsResponse: Codable {
    let success: Bool
    let data: [PlayerStats]
    let count: Int
    let filters: PlayerStatsFilters?
}

struct PlayerStatsFilters: Codable {
    let gameId: String?
    let playerId: String?
    let teamId: String?
}
